### PR TITLE
[Feature][Connector-V2] [Hudi]Add hudi sink connector

### PR DESCRIPTION
**Pull request by liugddx**
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by davidzollo** [comment]
@liugddx , is this feature ready now?  I accidentally opened 'ready for review'

**Comment by liugddx** [comment]
> @liugddx , is this feature ready now? I accidentally opened 'ready for review'

Sorry, I'm still learning hudi. i will finish it as soon as possible.

**Comment by Carl-Zhou-CN** [comment]
hi，@liugddx  Do I see any new considerations in removing hudi source?

**Comment by nbeeee** [comment]
@liugddx when the Hudi Sink can be used?

**Comment by CallStorm** [comment]
when the Hudi Sink can be used?

File: seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/source/HudiSource.java [hunk 1]
```diff
@@ -1,164 +0,0 @@
-/*
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/source/HudiSource.java:1]: Any reason for remove hudi source?

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-hudi-e2e/src/test/java/org/apache/seatunnel/e2e/connector/hudi/HudiIT.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        value = {TestContainerId.SPARK_2_4},
+        type = {},
+        disabledReason = "")
+public class HudiIT extends TestSuiteBase {
+
+    @TestTemplate
+    public void testWriteHudi(TestContainer container) throws IOException, InterruptedException {
+        Container.ExecResult textWriteResult = container.executeJob("/fake_to_hudi.conf");
+        Assertions.assertEquals(0, textWriteResult.getExitCode());
+    }
```
Comment (Hisoka-X) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-hudi-e2e/src/test/java/org/apache/seatunnel/e2e/connector/hudi/HudiIT.java]: Please check data in hudi to make sure the data already write into hudi.

File: seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/source/HudiSource.java (hunk 1 above)
Comment (liugddx) [reply seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/source/HudiSource.java:1]: According to the design principles of the SeaTunnel connector, it should not rely on a specific engine. Hudi source currently must rely on the specified engine. It does not have a Java SDK version, so delete Hudi source for now.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-hudi-e2e/src/test/java/org/apache/seatunnel/e2e/connector/hudi/HudiIT.java (hunk 2 above)
Comment (liugddx) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-hudi-e2e/src/test/java/org/apache/seatunnel/e2e/connector/hudi/HudiIT.java]: Done.

File: docs/en/connector-v2/sink/Hudi.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+> Hudi sink connector
+
+## Description
+
+Used to write data to Hudi.
+
+## Key features
+
+- [x] [exactly-once](../../concept/connector-v2-features.md)
+
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md:13]: `- [x] [cdc](../../concept/connector-v2-features.md)`

File: docs/en/connector-v2/sink/Hudi.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+| upsert_shuffle_parallelism | Int    | no       | 2             |
+| min_commits_to_keep        | Int    | no       | 20            |
+| max_commits_to_keep        | Int    | no       | 30            |
+| common-options             | config | no       | -             |
+
+### table_name [string]
+
+`table_name` The name of hudi table.
+
+### table_path [string]
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]: `table_dfs_path` is better. `table_path` is same as the `JDBC Connector` and  may confusion with `table_path` in other connectors.

File: docs/en/connector-v2/sink/Hudi.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+`table.path` The dfs root path of hudi table,such as 'hdfs://nameserivce/data/hudi/hudi_table/'.
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]: `table.path` ? it is not same as the name in `Options`

File: docs/en/connector-v2/sink/Hudi.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+### table_type [string]
+
+`table_type` The type of hudi table.
+
+### conf_files [string]
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]: `conf_files_path` is better.

File: docs/en/connector-v2/sink/Hudi.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+`conf_files` The environment conf file path list(local path), which used to init hdfs client to read hudi table file. The example is '/home/test/hdfs-site.xml;/home/test/core-site.xml;/home/test/yarn-site.xml'.
+
+### op_type [boolean]
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]:
```suggestion
### op_type [enum]
```

File: docs/en/connector-v2/source/Hudi.md [hunk 8]
```diff
@@ -1,90 +0,0 @@
-# Hudi
```
Comment (EricJoy2048) [inline docs/en/connector-v2/source/Hudi.md:1]: delete this file? An empty file is meaningless.

File: seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/config/HudiSinkConfig.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        builder.batchIntervalMs(
+                optionalBatchIntervalMs.orElseGet(HudiOptions.BATCH_INTERVAL_MS::defaultValue));
+
+        builder.partitionFields(
+                partitionFields.orElseGet(HudiOptions.PARTITION_FIELDS::defaultValue));
+
+        builder.recordKeyFields(
+                recordKeyFields.orElseGet(HudiOptions.RECORD_KEY_FIELDS::defaultValue));
+
+        builder.insertShuffleParallelism(
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/config/HudiSinkConfig.java]: why not use `builder.insertShuffleParallelism(config.get(HudiOptions.UPSERT_SHUFFLE_PARALLELISM))` ?

File: seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/sink/HudiSink.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.seatunnel.connectors.seatunnel.hudi.state.HudiSinkState;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class HudiSink
+        implements SeaTunnelSink<
+                SeaTunnelRow, HudiSinkState, HudiCommitInfo, HudiAggregatedCommitInfo> {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/sink/HudiSink.java]: Does this connector support `SupportMultiTableSink` ?

File: seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/sink/HudiSinkFactory.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+@AutoService(Factory.class)
+public class HudiSinkFactory implements TableSinkFactory {
+    @Override
+    public String factoryIdentifier() {
+        return "Hudi";
+    }
+
+    @Override
+    public OptionRule optionRule() {
+        return OptionRule.builder().required(TABLE_PATH, TABLE_NAME).build();
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/sink/HudiSinkFactory.java]: You lost some options here, can you add all options here which you put in the document?

File: docs/en/connector-v2/sink/Hudi.md [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+Source plugin common parameters, please refer to [Source Common Options](common-options.md) for details.
+
+## Examples
+
+```hocon
+source {
+
+  Hudi {
+    table.path = "hdfs://nameserivce/data/hudi/hudi_table/"
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]:
```suggestion
    table_dfs_path = "hdfs://nameserivce/data/hudi/hudi_table/"
```

File: docs/en/connector-v2/sink/Hudi.md [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    table.type = "cow"
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]:
```suggestion
    table_type = "cow"
```

File: docs/en/connector-v2/sink/Hudi.md [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+`table_name` The name of hudi table.
+
+### table_dfs_path [string]
+
+`table_dfs_path` The dfs root path of hudi table,such as 'hdfs://nameserivce/data/hudi/hudi_table/'.
+
+### table_type [string]
+
+`table_type` The type of hudi table.
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]: Does this option‘s type is `enum` ?

File: docs/en/connector-v2/sink/Hudi.md [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+## Examples
+
+```hocon
+source {
+
+  Hudi {
+    table.path = "hdfs://nameserivce/data/hudi/hudi_table/"
+    table.type = "cow"
+    conf.files = "/home/test/hdfs-site.xml;/home/test/core-site.xml;/home/test/yarn-site.xml"
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]: Please check the option keys in this example.

File: docs/en/connector-v2/sink/Hudi.md [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+### table_name [string]
+
+`table_name` The name of hudi table.
+
+### table_dfs_path [string]
+
+`table_dfs_path` The dfs root path of hudi table,such as 'hdfs://nameserivce/data/hudi/hudi_table/'.
+
+### table_type [string]
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hudi.md]:
```suggestion
### table_type [enum]
```

**Review by EricJoy2048 — APPROVED** [review]

**Review by hailin0 — APPROVED** [review]

File: seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/source/HudiSource.java (hunk 1 above)
Comment (dengc367) [reply seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/source/HudiSource.java:1]: what mean about must have a Java version?

**Comment by liugddx** [comment]
@leesf  PTAL

**Comment by sudohainguyen** [comment]
hey @liugddx, is there any plan to revive hudi source?

**Comment by liugddx** [comment]
The Hudi community does not have the source interface of Java sdk.So there is no plan for the time being.

**Pull request merged**